### PR TITLE
Add Custom metadata field to alias

### PR DESCRIPTION
**Pull request by akshya96**
Add custom_metadata field to alias 
Updated create, update, getbyaliasID and list 
Policy templating changes to include custom_metadata 

Testing :
curl --header X-Vault-Token:s.LoB6WLWugC7Y7qwKF1M49zO7 http://127.0.0.1:8200/v1/identity/entity-alias/id/12051e19-b256-0eed-35a0-879c2e308623
{"request_id":"459f2eec-5ded-da61-1237-da5c5d535712","lease_id":"","renewable":false,"lease_duration":0,"data":{"canonical_id":"da097a8e-224f-db02-9524-adfa2c48bb7d","creation_time":"2021-09-07T21:34:58.885245Z","custom_metadata":{"bar1":"1231","baz":"5c07d823-3810-48f6-a147-4c06b5219e84","foo1":"abc1"},"id":"12051e19-b256-0eed-35a0-879c2e308623","last_update_time":"2021-09-07T21:34:58.930261Z","merged_from_canonical_ids":null,"metadata":null,"mount_accessor":"auth_userpass_eadc45b0","mount_path":"auth/userpass/","mount_type":"userpass","name":"akshya2","namespace_id":"root"},"wrap_info":null,"warnings":null,"auth":null}

### CODE REVIEW

File: helper/identity/types.proto [hunk 1]
```diff
@@ -172,6 +172,11 @@ message Alias {
 	// NamespaceID is the identifier of the namespace to which this alias
 	// belongs.
 	string namespace_id = 11;
+
+	// Custom Metadata
+	map<string, string> customMetadata = 12;
+
```
Comment (vishalnayak) [inline helper/identity/types.proto:178]: Can we remove these newlines?

File: vault/identity_store_aliases.go [hunk 2]
```diff
@@ -44,6 +52,10 @@ This field is deprecated, use canonical_id.`,
 					Type:        framework.TypeString,
 					Description: "Name of the alias; unused for a modify",
 				},
+				"custom_metadata": {
+					Type:        framework.TypeKVPairs,
+					Description: "User-provided key-value pairs",
```
Comment (vishalnayak) [inline vault/identity_store_aliases.go:57]:
```suggestion
					Description: "User provided key-value pairs",
```

File: vault/identity_store_aliases.go [hunk 3]
```diff
@@ -77,6 +89,10 @@ This field is deprecated, use canonical_id.`,
 					Type:        framework.TypeString,
 					Description: "(Unused)",
 				},
+				"custom_metadata": {
+					Type:        framework.TypeKVPairs,
+					Description: "User-provided key-value pairs",
```
Comment (vishalnayak) [inline vault/identity_store_aliases.go:94]:
```suggestion
					Description: "User provided key-value pairs",
```

File: vault/identity_store_aliases.go [hunk 4]
```diff
@@ -143,31 +171,31 @@ func (i *IdentityStore) handleAliasCreateUpdate() framework.OperationFunc {
...
 				case name == "":
 					// No change to mount name
 					name = alias.Name
+				case len(customMetadata) == 0:
+					// No change to custom metadata
+					customMetadata = alias.CustomMetadata
 
 				default:
-					// Both provided
+					// mountAccessor,name and customMetadata  provided
```
Comment (vishalnayak) [inline vault/identity_store_aliases.go:196]:
```suggestion
					// mountAccessor, name and customMetadata  provided
```

File: vault/identity_store_aliases.go [hunk 5]
```diff
@@ -373,6 +403,56 @@ func (i *IdentityStore) handleAliasUpdate(ctx context.Context, req *logical.Requ
 	}, nil
 }
 
+//TODO: ask if we can use the same validation code from kv secrets plugin
```
Comment (vishalnayak) [inline vault/identity_store_aliases.go:406]: We can't for now. We can retain this function here.

**Review by vishalnayak — COMMENTED** [review]
Looking good!

File: sdk/logical/identity.proto [hunk 6]
```diff
@@ -50,6 +50,9 @@ message Alias {
 	// NamespaceID is the identifier of the namespace to which this alias
 	// belongs.
 	string namespace_id = 6;
+
+	// Custom Metadata represents the custom data tied to this entity
```
Comment (vishalnayak) [inline sdk/logical/identity.proto:54]:
```suggestion
	// Custom Metadata represents the custom data tied to this alias
```

File: helper/identity/types.proto [hunk 7]
```diff
@@ -172,6 +172,9 @@ message Alias {
 	// NamespaceID is the identifier of the namespace to which this alias
 	// belongs.
 	string namespace_id = 11;
+
+	// Custom Metadata
```
Comment (vishalnayak) [inline helper/identity/types.proto:176]:
```suggestion
	// Custom Metadata represents the custom data tied to this alias
```

**Review by vishalnayak — APPROVED** [review]

File: vault/identity_store_aliases.go [hunk 8]
```diff
@@ -296,16 +333,17 @@ func (i *IdentityStore) handleAliasUpdate(ctx context.Context, req *logical.Requ
 		if err != nil {
 			return nil, err
 		}
-		// Bail unless it's just a case change
-		if existingAlias != nil && !strings.EqualFold(existingAlias.Name, name) {
+
+		// Bail unless it's just a case change or no custom meta data change
+		if existingAlias != nil && !strings.EqualFold(existingAlias.Name, name) && reflect.DeepEqual(customMetadata, alias.CustomMetadata) {
```
Comment (ncabatoff) [inline vault/identity_store_aliases.go:338]: I don't think this check should be changed.  It's also going to collide with https://github.com/hashicorp/vault/pull/12473, which I'm about to merge.

File: vault/identity_store_aliases.go [hunk 9]
```diff
@@ -266,10 +303,10 @@ func (i *IdentityStore) handleAliasCreate(ctx context.Context, req *logical.Requ
 	}, nil
 }
 
-func (i *IdentityStore) handleAliasUpdate(ctx context.Context, req *logical.Request, canonicalID, name, mountAccessor string, alias *identity.Alias) (*logical.Response, error) {
+func (i *IdentityStore) handleAliasUpdate(ctx context.Context, req *logical.Request, canonicalID, name, mountAccessor string, alias *identity.Alias, customMetadata map[string]string) (*logical.Response, error) {
 	if name == alias.Name &&
 		mountAccessor == alias.MountAccessor &&
-		(canonicalID == alias.CanonicalID || canonicalID == "") {
+		(canonicalID == alias.CanonicalID || canonicalID == "") && (reflect.DeepEqual(customMetadata, alias.CustomMetadata)) {
```
Comment (vishalnayak) [inline vault/identity_store_aliases.go:309]: Do you want to try `strutil.EqualStringMaps` instead of `reflect.DeepEqual`?

Comment (vishalnayak) [reply vault/identity_store_aliases.go:309]: This comment is applicable to all places where `reflect.DeepEqual` is used.

File: helper/identity/types.proto (hunk 1 above)
Comment (akshya96) [reply helper/identity/types.proto]: done

File: vault/identity_store_aliases.go (hunk 2 above)
Comment (akshya96) [reply vault/identity_store_aliases.go]: done

File: vault/identity_store_aliases.go (hunk 3 above)
Comment (akshya96) [reply vault/identity_store_aliases.go]: done

File: vault/identity_store_aliases.go (hunk 5 above)
Comment (akshya96) [reply vault/identity_store_aliases.go]: done

File: sdk/logical/identity.proto (hunk 6 above)
Comment (akshya96) [reply sdk/logical/identity.proto]: done

File: helper/identity/types.proto (hunk 7 above)
Comment (akshya96) [reply helper/identity/types.proto]: done

File: vault/identity_store_aliases.go (hunk 8 above)
Comment (akshya96) [reply vault/identity_store_aliases.go]: done

File: vault/identity_store_aliases.go (hunk 9 above)
Comment (akshya96) [reply vault/identity_store_aliases.go]: done

File: vault/identity_store_aliases.go (hunk 4 above)
Comment (akshya96) [reply vault/identity_store_aliases.go]: done

**Pull request merged**